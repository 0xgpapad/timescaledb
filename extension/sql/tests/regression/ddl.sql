\set ON_ERROR_STOP 1
\set VERBOSITY verbose
\set SHOW_CONTEXT never

\o /dev/null
\ir include/create_clustered_db.sql

\o
\set ECHO ALL
\c meta
SELECT add_cluster_user('postgres', NULL);

SELECT set_meta('meta' :: NAME, 'localhost');
SELECT add_node('Test1' :: NAME, 'localhost');
SELECT add_node('test2' :: NAME, 'localhost');

\c Test1


\ir include/ddl_ops_1.sql

SELECT * FROM PUBLIC."Hypertable_1";
EXPLAIN SELECT * FROM PUBLIC."Hypertable_1";
SELECT * FROM ONLY PUBLIC."Hypertable_1";
EXPLAIN SELECT * FROM ONLY PUBLIC."Hypertable_1";

\d+ PUBLIC."Hypertable_1"
\d+ "_sysinternal"."_hyper_1_root"
\d+ _sysinternal._hyper_1_1_0_1_data
SELECT * FROM PUBLIC.default_replica_node;


\c test2

\d+ PUBLIC."Hypertable_1"
\d+ "_sysinternal"."_hyper_1_root"

\ir include/ddl_ops_2.sql

\d+ PUBLIC."Hypertable_1"
\d+ "_sysinternal"."_hyper_1_root"
SELECT * FROM _sysinternal._hyper_1_0_1_distinct_data;


\c Test1
\d+ PUBLIC."Hypertable_1"
\d+ "_sysinternal"."_hyper_1_root"
\d+ _sysinternal._hyper_1_1_0_1_data

SELECT * FROM PUBLIC."Hypertable_1";
