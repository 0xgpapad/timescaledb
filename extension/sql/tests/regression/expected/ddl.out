 setup_meta 
------------
 
(1 row)

 setup_main 
------------
 
(1 row)

 setup_main 
------------
 
(1 row)

\c meta
SELECT add_cluster_user('postgres', NULL);
 add_cluster_user 
------------------
 
(1 row)

SELECT set_meta('meta' :: NAME, 'localhost');
 set_meta 
----------
 
(1 row)

SELECT add_node('Test1' :: NAME, 'localhost');
 add_node 
----------
 
(1 row)

SELECT add_node('test2' :: NAME, 'localhost');
 add_node 
----------
 
(1 row)

\c Test1
CREATE TABLE PUBLIC."Hypertable_1" (
  time BIGINT NOT NULL,
  "Device_id" TEXT NOT NULL,
  temp_c int NOT NULL DEFAULT -1,
  humidity numeric NULL DEFAULT 0,
  sensor_1 NUMERIC NULL DEFAULT 1,
  sensor_2 NUMERIC NOT NULL DEFAULT 1,
  sensor_3 NUMERIC NOT NULL DEFAULT 1,
  sensor_4 NUMERIC NOT NULL DEFAULT 1
);
CREATE INDEX ON PUBLIC."Hypertable_1" (time, "Device_id");
SELECT * FROM add_hypertable('"public"."Hypertable_1"', 'time', 'Device_id');
         name          | main_schema_name | main_table_name | associated_schema_name | associated_table_prefix | root_schema_name | root_table_name | distinct_schema_name | distinct_table_name | replication_factor | placement | time_field_name | time_field_type | created_on 
-----------------------+------------------+-----------------+------------------------+-------------------------+------------------+-----------------+----------------------+---------------------+--------------------+-----------+-----------------+-----------------+------------
 public."Hypertable_1" | public           | Hypertable_1    | _sys_1_                | _hyper_1                | _sys_1_          | _hyper_1_root   | _sys_1_              | _hyper_1_distinct   |                  1 | STICKY    | time            | bigint          | Test1
(1 row)

SELECT * FROM hypertable;
         name          | main_schema_name | main_table_name | associated_schema_name | associated_table_prefix | root_schema_name | root_table_name | distinct_schema_name | distinct_table_name | replication_factor | placement | time_field_name | time_field_type | created_on 
-----------------------+------------------+-----------------+------------------------+-------------------------+------------------+-----------------+----------------------+---------------------+--------------------+-----------+-----------------+-----------------+------------
 public."Hypertable_1" | public           | Hypertable_1    | _sys_1_                | _hyper_1                | _sys_1_          | _hyper_1_root   | _sys_1_              | _hyper_1_distinct   |                  1 | STICKY    | time            | bigint          | Test1
(1 row)

SELECT * FROM hypertable_index;
    hypertable_name    | main_schema_name |         main_index_name         |                                   definition                                    | created_on 
-----------------------+------------------+---------------------------------+---------------------------------------------------------------------------------+------------
 public."Hypertable_1" | public           | Hypertable_1_time_Device_id_idx | CREATE INDEX /*INDEX_NAME*/ ON /*TABLE_NAME*/ USING btree ("time", "Device_id") | Test1
(1 row)

CREATE INDEX ON PUBLIC."Hypertable_1" (time, "temp_c");
CREATE INDEX "ind_humidity" ON PUBLIC."Hypertable_1" (time, "humidity");
CREATE INDEX "ind_sensor_1" ON PUBLIC."Hypertable_1" (time, "sensor_1");
SELECT set_is_distinct_flag('"public"."Hypertable_1"', 'sensor_1', TRUE);
 set_is_distinct_flag 
----------------------
 
(1 row)

SELECT set_is_distinct_flag('"public"."Hypertable_1"', 'sensor_2', TRUE);
 set_is_distinct_flag 
----------------------
 
(1 row)

INSERT INTO "Hypertable_1"(time, "Device_id", temp_c, humidity, sensor_1, sensor_2, sensor_3, sensor_4)
VALUES(1257894000000000000, 'dev1', 30, 70, 1, 2, 3, 100);
SELECT * FROM PUBLIC."Hypertable_1";
        time         | Device_id | temp_c | humidity | sensor_1 | sensor_2 | sensor_3 | sensor_4 
---------------------+-----------+--------+----------+----------+----------+----------+----------
 1257894000000000000 | dev1      |     30 |       70 |        1 |        2 |        3 |      100
(1 row)

EXPLAIN SELECT * FROM PUBLIC."Hypertable_1";
                                   QUERY PLAN                                    
---------------------------------------------------------------------------------
 Append  (cost=0.00..27.00 rows=702 width=204)
   ->  Seq Scan on _hyper_1_0_replica  (cost=0.00..0.00 rows=1 width=204)
   ->  Seq Scan on _hyper_1_1_0_partition  (cost=0.00..0.00 rows=1 width=204)
   ->  Seq Scan on _hyper_1_2_0_partition  (cost=0.00..13.50 rows=350 width=204)
   ->  Seq Scan on _hyper_1_1_0_1_data  (cost=0.00..13.50 rows=350 width=204)
(5 rows)

SELECT * FROM ONLY PUBLIC."Hypertable_1";
 time | Device_id | temp_c | humidity | sensor_1 | sensor_2 | sensor_3 | sensor_4 
------+-----------+--------+----------+----------+----------+----------+----------
(0 rows)

EXPLAIN SELECT * FROM ONLY PUBLIC."Hypertable_1";
                            QUERY PLAN                             
-------------------------------------------------------------------
 Seq Scan on "Hypertable_1"  (cost=0.00..13.50 rows=350 width=204)
(1 row)

\d+ PUBLIC."Hypertable_1"
                                 Table "public.Hypertable_1"
  Column   |  Type   |           Modifiers            | Storage  | Stats target | Description 
-----------+---------+--------------------------------+----------+--------------+-------------
 time      | bigint  | not null                       | plain    |              | 
 Device_id | text    | not null                       | extended |              | 
 temp_c    | integer | not null default '-1'::integer | plain    |              | 
 humidity  | numeric | default 0                      | main     |              | 
 sensor_1  | numeric | default 1                      | main     |              | 
 sensor_2  | numeric | not null default 1             | main     |              | 
 sensor_3  | numeric | not null default 1             | main     |              | 
 sensor_4  | numeric | not null default 1             | main     |              | 
Indexes:
    "Hypertable_1_time_Device_id_idx" btree ("time", "Device_id")
    "Hypertable_1_time_temp_c_idx" btree ("time", temp_c)
    "ind_humidity" btree ("time", humidity)
    "ind_sensor_1" btree ("time", sensor_1)
Triggers:
    insert_trigger AFTER INSERT ON "Hypertable_1" FOR EACH STATEMENT EXECUTE PROCEDURE _sysinternal.on_main_table_insert()

\d+ "_sys_1_"."_hyper_1_root"
                                Table "_sys_1_._hyper_1_root"
  Column   |  Type   |           Modifiers            | Storage  | Stats target | Description 
-----------+---------+--------------------------------+----------+--------------+-------------
 time      | bigint  | not null                       | plain    |              | 
 Device_id | text    | not null                       | extended |              | 
 temp_c    | integer | not null default '-1'::integer | plain    |              | 
 humidity  | numeric | default 0                      | main     |              | 
 sensor_1  | numeric | default 1                      | main     |              | 
 sensor_2  | numeric | not null default 1             | main     |              | 
 sensor_3  | numeric | not null default 1             | main     |              | 
 sensor_4  | numeric | not null default 1             | main     |              | 
Child tables: _sys_1_._hyper_1_0_replica

\d+ _sys_1_._hyper_1_1_0_1_data
                             Table "_sys_1_._hyper_1_1_0_1_data"
  Column   |  Type   |           Modifiers            | Storage  | Stats target | Description 
-----------+---------+--------------------------------+----------+--------------+-------------
 time      | bigint  | not null                       | plain    |              | 
 Device_id | text    | not null                       | extended |              | 
 temp_c    | integer | not null default '-1'::integer | plain    |              | 
 humidity  | numeric | default 0                      | main     |              | 
 sensor_1  | numeric | default 1                      | main     |              | 
 sensor_2  | numeric | not null default 1             | main     |              | 
 sensor_3  | numeric | not null default 1             | main     |              | 
 sensor_4  | numeric | not null default 1             | main     |              | 
Indexes:
    "1-Hypertable_1_time_Device_id_idx" btree ("time", "Device_id")
    "2-Hypertable_1_time_temp_c_idx" btree ("time", temp_c)
    "3-ind_humidity" btree ("time", humidity)
    "4-ind_sensor_1" btree ("time", sensor_1)
Check constraints:
    "partition" CHECK (get_partition_for_key("Device_id", 32768) >= '0'::smallint AND get_partition_for_key("Device_id", 32768) <= '16383'::smallint)
Inherits: _sys_1_._hyper_1_1_0_partition

SELECT * FROM PUBLIC.default_replica_node;
 database_name |    hypertable_name    | replica_id 
---------------+-----------------------+------------
 Test1         | public."Hypertable_1" |          0
 test2         | public."Hypertable_1" |          0
(2 rows)

\c test2
\d+ PUBLIC."Hypertable_1"
                                 Table "public.Hypertable_1"
  Column   |  Type   |           Modifiers            | Storage  | Stats target | Description 
-----------+---------+--------------------------------+----------+--------------+-------------
 time      | bigint  | not null                       | plain    |              | 
 Device_id | text    | not null                       | extended |              | 
 temp_c    | integer | not null default '-1'::integer | plain    |              | 
 humidity  | numeric | default 0                      | main     |              | 
 sensor_1  | numeric | default 1                      | main     |              | 
 sensor_2  | numeric | not null default 1             | main     |              | 
 sensor_3  | numeric | not null default 1             | main     |              | 
 sensor_4  | numeric | not null default 1             | main     |              | 
Indexes:
    "Hypertable_1_time_Device_id_idx" btree ("time", "Device_id")
    "Hypertable_1_time_temp_c_idx" btree ("time", temp_c)
    "ind_humidity" btree ("time", humidity)
    "ind_sensor_1" btree ("time", sensor_1)
Triggers:
    insert_trigger AFTER INSERT ON "Hypertable_1" FOR EACH STATEMENT EXECUTE PROCEDURE _sysinternal.on_main_table_insert()

\d+ "_sys_1_"."_hyper_1_root"
                                Table "_sys_1_._hyper_1_root"
  Column   |  Type   |           Modifiers            | Storage  | Stats target | Description 
-----------+---------+--------------------------------+----------+--------------+-------------
 time      | bigint  | not null                       | plain    |              | 
 Device_id | text    | not null                       | extended |              | 
 temp_c    | integer | not null default '-1'::integer | plain    |              | 
 humidity  | numeric | default 0                      | main     |              | 
 sensor_1  | numeric | default 1                      | main     |              | 
 sensor_2  | numeric | not null default 1             | main     |              | 
 sensor_3  | numeric | not null default 1             | main     |              | 
 sensor_4  | numeric | not null default 1             | main     |              | 
Child tables: _sys_1_._hyper_1_0_replica

SELECT set_is_distinct_flag('"public"."Hypertable_1"', 'sensor_2', FALSE);
 set_is_distinct_flag 
----------------------
 
(1 row)

SELECT set_is_distinct_flag('"public"."Hypertable_1"', 'Device_id', TRUE);
 set_is_distinct_flag 
----------------------
 
(1 row)

ALTER TABLE PUBLIC."Hypertable_1" ADD COLUMN temp_f INTEGER NOT NULL DEFAULT 31;
ALTER TABLE PUBLIC."Hypertable_1" DROP COLUMN temp_c;
ALTER TABLE PUBLIC."Hypertable_1" DROP COLUMN sensor_4;
ALTER TABLE PUBLIC."Hypertable_1" ALTER COLUMN humidity SET DEFAULT 100;
ALTER TABLE PUBLIC."Hypertable_1" ALTER COLUMN sensor_1 DROP DEFAULT;
ALTER TABLE PUBLIC."Hypertable_1" ALTER COLUMN sensor_2 SET DEFAULT NULL;
ALTER TABLE PUBLIC."Hypertable_1" ALTER COLUMN sensor_1 SET NOT NULL;
ALTER TABLE PUBLIC."Hypertable_1" ALTER COLUMN sensor_2 DROP NOT NULL;
ALTER TABLE PUBLIC."Hypertable_1" RENAME COLUMN sensor_2 TO sensor_2_renamed;
ALTER TABLE PUBLIC."Hypertable_1" RENAME COLUMN sensor_3 TO sensor_3_renamed;
DROP INDEX "ind_sensor_1";
--expect error cases
\set ON_ERROR_STOP 0
ALTER TABLE PUBLIC."Hypertable_1" ALTER COLUMN sensor_2_renamed SET DATA TYPE int;
ALTER INDEX "ind_humidity" RENAME TO "ind_humdity2";
\set ON_ERROR_STOP 1
--create column with same name as previously renamed one
ALTER TABLE PUBLIC."Hypertable_1" ADD COLUMN sensor_3 BIGINT NOT NULL DEFAULT 131;
--create column with same name as previously dropped one
ALTER TABLE PUBLIC."Hypertable_1" ADD COLUMN sensor_4 BIGINT NOT NULL DEFAULT 131;
\d+ PUBLIC."Hypertable_1"
                                Table "public.Hypertable_1"
      Column      |  Type   |      Modifiers       | Storage  | Stats target | Description 
------------------+---------+----------------------+----------+--------------+-------------
 time             | bigint  | not null             | plain    |              | 
 Device_id        | text    | not null             | extended |              | 
 humidity         | numeric | default 100          | main     |              | 
 sensor_1         | numeric | not null             | main     |              | 
 sensor_2_renamed | numeric |                      | main     |              | 
 sensor_3_renamed | numeric | not null default 1   | main     |              | 
 temp_f           | integer | not null default 31  | plain    |              | 
 sensor_3         | bigint  | not null default 131 | plain    |              | 
 sensor_4         | bigint  | not null default 131 | plain    |              | 
Indexes:
    "Hypertable_1_time_Device_id_idx" btree ("time", "Device_id")
    "ind_humidity" btree ("time", humidity)
Triggers:
    insert_trigger AFTER INSERT ON "Hypertable_1" FOR EACH STATEMENT EXECUTE PROCEDURE _sysinternal.on_main_table_insert()

\d+ "_sys_1_"."_hyper_1_root"
                                Table "_sys_1_._hyper_1_root"
      Column      |  Type   |       Modifiers        | Storage  | Stats target | Description 
------------------+---------+------------------------+----------+--------------+-------------
 time             | bigint  | not null               | plain    |              | 
 Device_id        | text    | not null               | extended |              | 
 humidity         | numeric | default '100'::numeric | main     |              | 
 sensor_1         | numeric | not null               | main     |              | 
 sensor_2_renamed | numeric |                        | main     |              | 
 sensor_3_renamed | numeric | not null default 1     | main     |              | 
 temp_f           | integer | not null default 31    | plain    |              | 
 sensor_3         | bigint  | not null default 131   | plain    |              | 
 sensor_4         | bigint  | not null default 131   | plain    |              | 
Child tables: _sys_1_._hyper_1_0_replica

SELECT * FROM _sys_1_._hyper_1_0_1_distinct_data;
   field   | value 
-----------+-------
 sensor_1  | 1
 Device_id | dev1
(2 rows)

\c Test1
\d+ PUBLIC."Hypertable_1"
                                 Table "public.Hypertable_1"
      Column      |  Type   |       Modifiers        | Storage  | Stats target | Description 
------------------+---------+------------------------+----------+--------------+-------------
 time             | bigint  | not null               | plain    |              | 
 Device_id        | text    | not null               | extended |              | 
 humidity         | numeric | default '100'::numeric | main     |              | 
 sensor_1         | numeric | not null               | main     |              | 
 sensor_2_renamed | numeric |                        | main     |              | 
 sensor_3_renamed | numeric | not null default 1     | main     |              | 
 temp_f           | integer | not null default 31    | plain    |              | 
 sensor_3         | bigint  | not null default 131   | plain    |              | 
 sensor_4         | bigint  | not null default 131   | plain    |              | 
Indexes:
    "Hypertable_1_time_Device_id_idx" btree ("time", "Device_id")
    "ind_humidity" btree ("time", humidity)
Triggers:
    insert_trigger AFTER INSERT ON "Hypertable_1" FOR EACH STATEMENT EXECUTE PROCEDURE _sysinternal.on_main_table_insert()

\d+ "_sys_1_"."_hyper_1_root"
                                Table "_sys_1_._hyper_1_root"
      Column      |  Type   |       Modifiers        | Storage  | Stats target | Description 
------------------+---------+------------------------+----------+--------------+-------------
 time             | bigint  | not null               | plain    |              | 
 Device_id        | text    | not null               | extended |              | 
 humidity         | numeric | default '100'::numeric | main     |              | 
 sensor_1         | numeric | not null               | main     |              | 
 sensor_2_renamed | numeric |                        | main     |              | 
 sensor_3_renamed | numeric | not null default 1     | main     |              | 
 temp_f           | integer | not null default 31    | plain    |              | 
 sensor_3         | bigint  | not null default 131   | plain    |              | 
 sensor_4         | bigint  | not null default 131   | plain    |              | 
Child tables: _sys_1_._hyper_1_0_replica

\d+ _sys_1_._hyper_1_1_0_1_data
                             Table "_sys_1_._hyper_1_1_0_1_data"
      Column      |  Type   |       Modifiers        | Storage  | Stats target | Description 
------------------+---------+------------------------+----------+--------------+-------------
 time             | bigint  | not null               | plain    |              | 
 Device_id        | text    | not null               | extended |              | 
 humidity         | numeric | default '100'::numeric | main     |              | 
 sensor_1         | numeric | not null               | main     |              | 
 sensor_2_renamed | numeric |                        | main     |              | 
 sensor_3_renamed | numeric | not null default 1     | main     |              | 
 temp_f           | integer | not null default 31    | plain    |              | 
 sensor_3         | bigint  | not null default 131   | plain    |              | 
 sensor_4         | bigint  | not null default 131   | plain    |              | 
Indexes:
    "1-Hypertable_1_time_Device_id_idx" btree ("time", "Device_id")
    "3-ind_humidity" btree ("time", humidity)
Check constraints:
    "partition" CHECK (get_partition_for_key("Device_id", 32768) >= '0'::smallint AND get_partition_for_key("Device_id", 32768) <= '16383'::smallint)
Inherits: _sys_1_._hyper_1_1_0_partition

SELECT * FROM PUBLIC."Hypertable_1";
        time         | Device_id | humidity | sensor_1 | sensor_2_renamed | sensor_3_renamed | temp_f | sensor_3 | sensor_4 
---------------------+-----------+----------+----------+------------------+------------------+--------+----------+----------
 1257894000000000000 | dev1      |       70 |        1 |                2 |                3 |     31 |      131 |      131
(1 row)

